component/discovery/transformer/transformer.go

```diff
@@ -0,0 +1,265 @@
+package transformer
+
+import (
+	"context"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/grafana/agent/component"
+	"github.com/grafana/regexp"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+	"github.com/prometheus/prometheus/scrape"
+	"github.com/rfratto/gohcl"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "discovery.transformer",
+		Args:    Arguments{},
+		Exports: Exports{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the discovery.transformer component.
+type Arguments struct {
+	// Targets contains the input 'targets' passed by a Service Discovery component.
+	Targets []Target `hcl:"targets,optional"`
+
+	// The following values can be used to modify the original target label set.
+	RelabelConfigs []*RelabelConfig `hcl:"relabel_configs,block"`
+	ScrapeLabels   ScrapeLabels     `hcl:"scrape_labels,optional"`
+	QueryParams    url.Values       `hcl:"query_params,optional"`
+}
+
+// Target refers to a singular HTTP or HTTPS endpoint that will be used for scraping.
+// It mimics github.com/prometheus/prometheus/scrape.Target.
+// Here, we're using a map[string]string instead of labels.Labels; if the label ordering
+// is important, we can change to follow the upstream logic instead.
+// TODO (@tpaschalis) Maybe the target definitions should be part of the
+// Service Discovery components package. Let's reconsider once it's ready.
+type Target map[string]string
```

Comment:
I think there's two types of Targets we will be dealing with: 


1. Service Discovery targets contain a label set which identify some identified service. They may have common labels like `__address__`. 
2. Scraping targets encapsulate a service discovery target and add extra information like query parameters. 

I'm not sure we need to be dealing with the latter, since the scrape targets are an internal concept to the scrape manager (and are only exposed for the scrape targets API). 

We can likely remove all code related to scraping and just focus on relabeling target sets. WDYT?

Reply:
Yeah, I agree. To provide some context, initially the extra scrape labels and query params were to be able to call the Prometheus `scrape.PopulateLabels` method. 

Since this is a higher-level component, it makes sense to only deal with the discovered label set and keep it generic enough; these scrape-specific details can be handled by a downstream component. I'll make this change and get back to you.

Reply:
To provide more information: a list of things that [scrape.PopulateLabels](https://github.com/prometheus/prometheus/blob/867d3bd78f514ffffcfd5da751cb2bd84c6aba72/scrape/target.go#L354) does when dealing with targets:

* Override scrape-specific parameters if they weren't set
* Encode query parameters as labels
* Drop the target if no __address__ is set
* Perform all relabelling steps
* Check if we should add a default port to the address
* Check that the address doesn't contain any url paths (no '/')
* Set scrape intervals and scrape timeout, ensure the timeout is no longer than the interval
* If the instace label isn't set, default it to the value of __address__
* Remove `__meta*` labels
* Checks if all label values are valid UTF8 strings

I've simplified the logic keeping only two parts. Applying the relabelling rules, and dropping the target if there's no __address__ or if a label is not a valid UTF8 string. Is that more in line with what you were thinking?

Reply:
I think you only need to call relabel.Process, and leave any processing of the labels alone :) 

All the rules you mentioned seem specific to the implementation of the scraper, and so we'd probably want to leave that to the scraper to handle.

Reply:
Gotcha! Changed the implementation to only run relabelling rules and just drop the target if the result labelset is empty.